### PR TITLE
server: fix registration redirect for non-root URLs

### LINKED ISSUE #553: Registration handler doesn't take full issuer URL into account
Here is the problem in handleAuthFunc() function:

```
if register {
	_, ok := idpc.(*connector.LocalConnector)
	if ok {
		q := url.Values{}
		q.Set("code", key)
		ru := httpPathRegister + "?" + q.Encode()
		w.Header().Set("Location", ru)
		w.WriteHeader(http.StatusFound)
		return
	}
}
```

It always redirects to httpPathRegister ("/register") w/o using issuerURL. So it's impossible to register a new user if dex is listening on a non-root path.

### PR DESCRIPTION
**Pull request by ericchiang**
Closes #553

cc @sym3tri @chancez

### CODE REVIEW

**Comment by ericchiang** [comment]
there's even a test!

File: server/http.go [hunk 1]
```diff
@@ -358,7 +358,7 @@ func handleAuthFunc(srv OIDCServer, idpcs []connector.Connector, tpl *template.T
 			if ok {
 				q := url.Values{}
 				q.Set("code", key)
-				ru := httpPathRegister + "?" + q.Encode()
+				ru := strings.TrimRight(baseURL, "/") + "/" + strings.TrimLeft(httpPathRegister, "/") + "?" + q.Encode()
```
Comment (chancez) [inline server/http.go]: Could you maybe use `path.Join` or `url.Parse` instead? It should remove some of the need to trim, but is also more likely work correctly with joining paths together.

Comment (ericchiang) [inline server/http.go]: Sure, I'll switch it to passing a URL and use path.Join

Comment (chancez) [inline server/http.go]: Actually, it seems weird that you're having to pass the basePath through to so many _different_ places. Is this a thing you only need to do here due to needing to set the `Location` header?

Comment (ericchiang) [inline server/http.go]: Yes.

**Comment by ericchiang** [comment]
updated

**Comment by ericchiang** [comment]
woops, I have no idea what just happened there. one sec

**Comment by ericchiang** [comment]
cc @chancez ready for another review

File: server/http_test.go [hunk 2]
```diff
@@ -210,6 +226,33 @@ func TestHandleAuthFuncResponsesSingleRedirectURL(t *testing.T) {
...
+			query: url.Values{
+				"response_type": []string{"code"},
+				"redirect_uri":  []string{"http://client.example.com/callback"},
+				"client_id":     []string{"client.example.com"},
+				"connector_id":  []string{"local"},
+				"register":      []string{"1"},
+				"scope":         []string{"openid"},
+			},
+			wantCode:     http.StatusFound,
+			wantLocation: "/register?code=code-2",
```
Comment (chancez) [inline server/http_test.go]: perhaps explicitly set baseURL to `/` here?

**Comment by chancez** [comment]
LGTM after nits.

**Pull request merged**